src/main/java/org/rumbledb/runtime/functions/sequences/general/SubsequenceFunctionIterator.java

```diff
@@ -50,107 +50,107 @@ public SubsequenceFunctionIterator(
+    protected void openLocal() {
+        int currentPosition = 1; // JSONiq indices start from 1
+
+        this.currentLength = this.length;
+        // first, perform all parameter checks (above)
+        // if length is 0, just return empty sequence
+        if (this.currentLength == 0) {
+            this.hasNext = false;
+            return;
+        } else {
+            this.sequenceIterator.open(this.currentDynamicContextForLocalExecution);
+
+            // find the start of the subsequence
+            while (this.sequenceIterator.hasNext()) {
+                if (currentPosition < this.startPosition) {
+                    this.sequenceIterator.next(); // skip item
+                } else {
+                    this.nextResult = this.sequenceIterator.next();
+                    // if length is specified, decrement it
+                    if (this.currentLength != -1) {
+                        this.currentLength--;
+                    }
+                    break;
+                }
+                currentPosition++;
             }
         }
 
-        // process start position param
-        RuntimeIterator positionIterator = this.children.get(1);
-        positionIterator.open(context);
-        if (!positionIterator.hasNext()) {
-            throw new UnexpectedTypeException(
-                    "Invalid args. subsequence can't be performed with empty sequence as the position",
-                    getMetadata()
-            );
-        }
-        Item positionItem = positionIterator.next();
-        if (positionItem.isArray()) {
-            throw new NonAtomicKeyException(
-                    "Invalid args. subsequence can't be performed with an array parameter as the position",
-                    getMetadata()
-            );
-        } else if (positionItem instanceof ObjectItem) {
-            throw new NonAtomicKeyException(
-                    "Invalid args. subsequence can't be performed with an object parameter as the position",
-                    getMetadata()
-            );
-        } else if (!(positionItem.isNumeric())) {
-            throw new UnexpectedTypeException(
-                    "Invalid args. Position parameter should be a numeric(Integer/Decimal/Double)",
-                    getMetadata()
-            );
+        // if startPosition overshoots, return empty sequence
+        if (this.nextResult == null) {
+            this.hasNext = false;
+            this.sequenceIterator.close();
+        } else {
+            this.hasNext = true;
         }
-        positionIterator.close();
-        // round double to nearest int
-        this.startPosition = (int) Math.round((positionItem.castToDoubleValue()));
+    }
+
+    @Override
+    protected void closeLocal() {
+        this.sequenceIterator.close();
+    }
 
+    @Override
+    protected void resetLocal(DynamicContext context) {
+        int currentPosition = 1; // JSONiq indices start from 1
+
+        this.currentLength = this.length;
```

Comment:
in the future, the code in reset and open, which is duplicate, could be factored out as an init method.

Reply:
I have fixed this in the latest version